### PR TITLE
ssl: Refactor to avoid re-encoding of decoded certs

### PR DESCRIPTION
**Pull request by IngelaAndin**
Let public key define the record #cert{} that can be used to provide
the original DER certs received from the peer and the decoded version
together as input to certificate handling functions such as
public:pkix_path_validation/3
    
It will be more efficient as we can avoid conversions and also a way for
applications such as ssl to tolerate certs with some ASN1-encoding errors that can be worked around.

### CODE REVIEW

File: lib/ssl/src/ssl_certificate.erl [hunk 1]
```diff
@@ -103,7 +104,7 @@ trusted_cert_and_paths([Peer],  CertDbHandle, CertDbRef, PartialChainHandler) ->
     Chain = [#cert{der=Peer, otp=OtpCert}],
     case public_key:pkix_is_self_signed(OtpCert) of
         true ->
-            [{selfsigned_peer, [Peer]}];
+            [{selfsigned_peer, [#cert{der=Peer, otp=OtpCert}]}];
```
Comment (dgud) [inline lib/ssl/src/ssl_certificate.erl:107]: Use Chain created two lines above

Comment (IngelaAndin) [reply lib/ssl/src/ssl_certificate.erl:107]: Will do

File: lib/public_key/include/public_key.hrl [hunk 2]
```diff
@@ -78,6 +78,8 @@
 	  point
 	 }).
 
+-record(cert, {der, otp}).
+
 -define(unspecified, 0).
 -define(keyCompromise, 1).
 -define(cACompromise, 2).
```
Comment (KennethL) [inline lib/public_key/include/public_key.hrl:85]: Why cACompromise and not caCompromise?

Comment (IngelaAndin) [reply lib/public_key/include/public_key.hrl:85]: These are enumeration names from an RFC and defined in an PKIX ASN-1 spec, and also not part of my PR!

**Comment by IngelaAndin** [comment]
Closing this to replace it with a more appropriate branch name.

**Pull request closed without merging**